app/models/auth/voms.rb

```diff
@@ -0,0 +1,51 @@
+require 'digest'
+
+module Auth
+  class Voms
+    class << self
+      def unified_credentials(hash)
+        dn = parse_hash_dn!(hash)
+        UnifiedCredentials.new(id: Digest::SHA256.hexdigest(dn),
+                               email: Rails.configuration.keystorm['default_email'],
+                               groups: parse_hash_groups!(hash),
+                               authentication: 'federation',
+                               name: parse_dn_name!(dn),
+                               identity: dn,
+                               expiration: parse_hash_exp!(hash))
+      end
+
+      private
+
+      def parse_hash_dn!(hash)
+        x509cred = hash.select { |key, value| /GRST_CRED_\d+/ =~ key && value.start_with?('X509USER') }
+        raise Errors::AuthError, 'voms hash has invalid X509USER "GRST_CRED_*" variable set' unless x509cred.size == 1
+        parsed = x509cred.values.first.split(' ', 5)
+        raise Errors::AuthError, 'failed to parse DN from voms hash' unless parsed.size == 5
+        parsed[4]
+      end
+
+      def parse_dn_name!(dn)
+        matched = /CN=([\w\s]*)$/.match(dn)
```

Comment:
I know ... but it would have to be matched by a way more complicated `Regexp`. Just ask Perun devs next door. So we have two choices here ... either we can very carefully match the LAST `CN=` and capture everything 'til the end of the line with `(.+)$` or just use the whole DN.